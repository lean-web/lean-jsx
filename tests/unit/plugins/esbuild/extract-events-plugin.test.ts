import { describe, test } from "@jest/globals";
import { HandlersPluginTester } from "./plugin-tester";
import fs from "fs";
import path from "path";

describe("extract-events-plugin", () => {
  test("test plugin", async () => {
    const { handlersCode, componentsCode, writtenFiles } =
      await new HandlersPluginTester().testExtractEventsPlugin({
        "file.tsx": `
    export function MyComponent() {
      return <button onclick={(ev) => console.log(ev)}>Click</button> 
    }
      `,
        "file2.tsx": `
        export function MyComponent2() {
            return <div>Hello</div>
        }
      `,
        "file3.tsx": `
        export function MyComponent3() {
            return <button onclick={(c1) => c1} onchange={(c) => c} onload={(c) => c}>Click</button> 
          }
      `,
      });
    expect(fs.promises.readFile).toHaveBeenCalledWith(
      "file.tsx",
      expect.any(String),
    );

    if (!handlersCode || !handlersCode.outputFiles) {
      throw new Error("Missing esbuild result");
    }

    expect(handlersCode.outputFiles[0].text).toMatchSnapshot();

    expect(
      Object.entries(writtenFiles).map(([fpath, contents]) => [
        path.relative(process.cwd(), fpath),
        contents,
      ]),
    ).toMatchSnapshot();

    expect(componentsCode.map((c) => c.contents).join("\n")).toMatchSnapshot();
  });

  test("test plugin - simulate rebuild", async () => {
    const tester = new HandlersPluginTester();
    const { handlersCode, componentsCode, writtenFiles } =
      await tester.testExtractEventsPlugin({
        "file.tsx": `
    export function MyComponent() {
      return <button onclick={(ev) => console.log(ev)}>Click</button> 
    }
      `,
        "file2.tsx": `
        export function MyComponent2() {
            return <div>Hello</div>
        }
      `,
      });
    expect(fs.promises.readFile).toHaveBeenCalledWith(
      "file.tsx",
      expect.any(String),
    );

    if (!handlersCode || !handlersCode.outputFiles) {
      throw new Error("Missing esbuild result");
    }

    expect(handlersCode.outputFiles?.map((f) => f.text)).toMatchSnapshot();

    expect(
      Object.entries(writtenFiles).map(([fpath, contents]) => [
        path.relative(process.cwd(), fpath),
        contents,
      ]),
    ).toMatchSnapshot();

    expect(componentsCode.map((c) => c.contents).join("\n")).toMatchSnapshot();

    // rebuild:

    const rebuildResults = await tester.testExtractEventsPlugin({
      "file.tsx": `
    export function MyComponent() {
      return <button onclick={(ev1) => console.log(ev1)}>Click</button> 
    }`,
    });

    expect(
      rebuildResults.handlersCode?.outputFiles?.map((f) => f.text),
    ).toMatchSnapshot("rebuild handlers code");

    expect(
      Object.entries(writtenFiles).map(([fpath, contents]) => [
        path.relative(process.cwd(), fpath),
        contents,
      ]),
    ).toMatchSnapshot("rebuild written files");

    expect(componentsCode.map((c) => c.contents).join("\n")).toMatchSnapshot(
      "rebuild components code",
    );
  });
});
