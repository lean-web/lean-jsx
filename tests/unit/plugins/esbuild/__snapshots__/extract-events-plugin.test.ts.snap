// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract-events-plugin test plugin - simulate rebuild 1`] = `
[
  "var sxl_user = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // <stdin>
  var stdin_exports = {};
  __export(stdin_exports, {
    fn14533fb3b06a348bf24a9f456f6b9275: () => fn14533fb3b06a348bf24a9f456f6b9275
  });
  var fn14533fb3b06a348bf24a9f456f6b9275 = (ev) => console.log(ev);
  return __toCommonJS(stdin_exports);
})();
",
]
`;

exports[`extract-events-plugin test plugin - simulate rebuild 2`] = `
[
  [
    ".lean/handlers/file.ts",
    "
import { withClientData } from "lean-web-utils/lib/server";
type Arg = Parameters<typeof withClientData>[1]
// file.tsx
export const fn14533fb3b06a348bf24a9f456f6b9275 = (ev) => console.log(ev);
",
  ],
]
`;

exports[`extract-events-plugin test plugin - simulate rebuild 3`] = `
"export function MyComponent() {
    return <button onclick={(ev) => { return window.sxl_user.fn14533fb3b06a348bf24a9f456f6b9275(ev); }}>Click</button>;
}

export function MyComponent2() {
    return <div>Hello</div>;
}
"
`;

exports[`extract-events-plugin test plugin - simulate rebuild: rebuild components code 1`] = `
"export function MyComponent() {
    return <button onclick={(ev) => { return window.sxl_user.fn14533fb3b06a348bf24a9f456f6b9275(ev); }}>Click</button>;
}

export function MyComponent2() {
    return <div>Hello</div>;
}
"
`;

exports[`extract-events-plugin test plugin - simulate rebuild: rebuild handlers code 1`] = `
[
  "var sxl_user = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // <stdin>
  var stdin_exports = {};
  __export(stdin_exports, {
    fnacbebc5075e7518791459a102562cdfc: () => fnacbebc5075e7518791459a102562cdfc
  });
  var fnacbebc5075e7518791459a102562cdfc = (ev1) => console.log(ev1);
  return __toCommonJS(stdin_exports);
})();
",
]
`;

exports[`extract-events-plugin test plugin - simulate rebuild: rebuild written files 1`] = `
[
  [
    ".lean/handlers/file.ts",
    "
import { withClientData } from "lean-web-utils/lib/server";
type Arg = Parameters<typeof withClientData>[1]
// file.tsx
export const fnacbebc5075e7518791459a102562cdfc = (ev1) => console.log(ev1);
",
  ],
]
`;

exports[`extract-events-plugin test plugin 1`] = `
"var sxl_user = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // <stdin>
  var stdin_exports = {};
  __export(stdin_exports, {
    fn14533fb3b06a348bf24a9f456f6b9275: () => fn14533fb3b06a348bf24a9f456f6b9275,
    fn65934597b0f8eb9debfc0762c97da7a6: () => fn65934597b0f8eb9debfc0762c97da7a6,
    fnc811db883db9743ff5764cd85fb9d030: () => fnc811db883db9743ff5764cd85fb9d030
  });
  var fn14533fb3b06a348bf24a9f456f6b9275 = (ev) => console.log(ev);
  var fnc811db883db9743ff5764cd85fb9d030 = (c1) => c1;
  var fn65934597b0f8eb9debfc0762c97da7a6 = (c) => c;
  return __toCommonJS(stdin_exports);
})();
"
`;

exports[`extract-events-plugin test plugin 2`] = `
[
  [
    ".lean/handlers/file.ts",
    "
import { withClientData } from "lean-web-utils/lib/server";
type Arg = Parameters<typeof withClientData>[1]
// file.tsx
export const fn14533fb3b06a348bf24a9f456f6b9275 = (ev) => console.log(ev);
",
  ],
  [
    ".lean/handlers/file3.ts",
    "
import { withClientData } from "lean-web-utils/lib/server";
type Arg = Parameters<typeof withClientData>[1]
// file3.tsx
export const fnc811db883db9743ff5764cd85fb9d030 = (c1) => c1;
export const fn65934597b0f8eb9debfc0762c97da7a6 = (c) => c;
",
  ],
]
`;

exports[`extract-events-plugin test plugin 3`] = `
"export function MyComponent() {
    return <button onclick={(ev) => { return window.sxl_user.fn14533fb3b06a348bf24a9f456f6b9275(ev); }}>Click</button>;
}

export function MyComponent2() {
    return <div>Hello</div>;
}

export function MyComponent3() {
    return <button onclick={(c1) => { return window.sxl_user.fnc811db883db9743ff5764cd85fb9d030(c1); }} onchange={(c) => { return window.sxl_user.fn65934597b0f8eb9debfc0762c97da7a6(c); }} onload={(c) => { return window.sxl_user.fn65934597b0f8eb9debfc0762c97da7a6(c); }}>Click</button>;
}
"
`;
